# Data Cleaning (Customer_Service Dataset)

In [2]:
# import necessory libraries

import numpy as np
import pandas as pd
import datetime as datetime
import matplotlib.pyplot as plt
import seaborn as sns
import random as random

In [224]:
# load dataset in jupitor notebook

data = pd.read_csv("C:\\Users\\vaidi\\Desktop\\Call_Center.csv")

In [225]:
# let's check our dataset

data.head()

id       customer_name      sentiment  csat_score  \
0  DKK-57076809-w-055481-fU    Analise Gairdner        Neutral         7.0   
1  QGK-72219678-w-102139-KY    Crichton Kidsley  Very Positive         NaN   
2  GYJ-30025932-A-023015-LD   Averill Brundrett       Negative         NaN   
3  ZJI-96807559-i-620008-m7     Noreen Lafflina  Very Negative         1.0   
4  DDU-69451719-O-176482-Fm  Toma Van der Beken  Very Positive         NaN   

  call_timestamp            reason         city           state      channel  \
0     10/29/2020  Billing Question      Detroit        Michigan  Call-Center   
1     10/05/2020    Service Outage  Spartanburg  South Carolina      Chatbot   
2     10/04/2020  Billing Question  Gainesville         Florida  Call-Center   
3     10/17/2020  Billing Question     Portland          Oregon      Chatbot   
4     10/17/2020          Payments   Fort Wayne         Indiana  Call-Center   

  response_time  call duration in minutes     call_center  
0    Within SLA                        17  Los Angeles/CA  
1    Within SLA                        23    Baltimore/MD  
2     Above SLA                        45  Los Angeles/CA  
3    Within SLA                        12  Los Angeles/CA  
4    Within SLA                        23  Los Angeles/CA

In [226]:
# Get information about the dataset

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32941 entries, 0 to 32940
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        32941 non-null  object 
 1   customer_name             32941 non-null  object 
 2   sentiment                 32941 non-null  object 
 3   csat_score                12271 non-null  float64
 4   call_timestamp            32941 non-null  object 
 5   reason                    32941 non-null  object 
 6   city                      32941 non-null  object 
 7   state                     32941 non-null  object 
 8   channel                   32941 non-null  object 
 9   response_time             32941 non-null  object 
 10  call duration in minutes  32941 non-null  int64  
 11  call_center               32941 non-null  object 
dtypes: float64(1), int64(1), object(10)
memory usage: 3.0+ MB


#### In this dataset :

* We have 12 fields and 32941 records.
* csat_score has missing value and has float64 datatype.
* call_timestamp field contains dates but has object datatype.

# A). Data Cleaning

### 1). Missing Values

In [227]:
# number of missing records in each column

miss_values = data.isna().sum()
miss_percent = round((data.isna().sum()/data.shape[0])*100)

print(miss_values, '\n\n', miss_percent)

id                              0
customer_name                   0
sentiment                       0
csat_score                  20670
call_timestamp                  0
reason                          0
city                            0
state                           0
channel                         0
response_time                   0
call duration in minutes        0
call_center                     0
dtype: int64 

 id                           0.0
customer_name                0.0
sentiment                    0.0
csat_score                  63.0
call_timestamp               0.0
reason                       0.0
city                         0.0
state                        0.0
channel                      0.0
response_time                0.0
call duration in minutes     0.0
call_center                  0.0
dtype: float64


* Only csat_score column has missing records but the number is huge 20,670 and in percentage terms 63% data is missing, which is huge and usually we drop any field with missing records more than 15%. But this is dummy data and it's a very important column in customer service alysis, so we will add some random records inplace of missing data.


In [21]:
data['csat_score'].head()

0    7.0
1    NaN
2    NaN
3    1.0
4    NaN
Name: csat_score, dtype: float64

* First convert the datatype from float64 to int, because we have to add random integer value. And, it's not possible to convert loat value to integer when it has "na, nan etc." values. So, first replace nan value with 0.

In [31]:
# replace nan value with 0

data['csat_score'].replace(np.nan, 0, inplace = True)

In [43]:
# now convert float datatype to int

data['csat_score'] = data['csat_score'].astype(int)

In [46]:
# Add random csat_score inplace of 0.

data['csat_score'] = data['csat_score'].apply(lambda x: np.random.randint(1,10) if x == 0 else x)

In [54]:
# let's see the final output

data['csat_score']

0        7
1        3
2        4
3        1
4        9
        ..
32936    8
32937    1
32938    3
32939    8
32940    2
Name: csat_score, Length: 32941, dtype: int64

* It is cleaned now and we have 32941 records.

###  2). Duplicate Records

In [162]:
# Check Duplicate Record in a dataset

data.duplicated().value_counts()

False    32941
dtype: int64

* All the records are unique in this datasest

### 3). Change Datatype

In this dataset only 2 fields required change.
* csat_score : we have already changed it.
* call_timestamp : let's convert it.

In [65]:
# converting call_timestamp from object to datetime dataype because it is not a string but a date field.

data['call_timestamp'] = data['call_timestamp'].astype('datetime64[ns]')

### 4). Rename Columns

Standardizig column name is necessory for clarity and easiry work flow.
so, we will convert
* call_timestamp : data
* call duration in minutes : call_duration

In [74]:
# Renaming field names

data.rename(columns = {'call_timestamp' : 'date', 'call duration in minutes' : 'call_duration'}, inplace = True)

### 5). Eliminate Leading and Trailing Space

Sometimes, text fields contais extra spaces before and after, so it's a good practice to make sure that text fields is clean.

First we will split our dataframe to two dataframe text_df and non_text_df. So, that we can remove the extra spaces in all the text columns and then join it to other column of the dataframe to get complete dataframe.

In [90]:
# create dataframe with only text objects

text_df = data.select_dtypes(include = 'object')

In [91]:
# create dataframe with all fields exept text

non_text_df = data.select_dtypes(exclude = 'object')

In [92]:
# remove leading and trailing spaces in all the text fiels

text_df = text_df.apply(lambda x: x.str.strip())

In [184]:
# assign text_df to new variable so that it will make more sense becasue now we will combine all fields

data2 = text_df

In [185]:
data2 = data2.join(non_text_df)

In [186]:
# now we have cleaned text columns in data2 dataframe with all fields. lets check it.

data2.head()

id       customer_name      sentiment  \
0  DKK-57076809-w-055481-fU    Analise Gairdner        Neutral   
1  QGK-72219678-w-102139-KY    Crichton Kidsley  Very Positive   
2  GYJ-30025932-A-023015-LD   Averill Brundrett       Negative   
3  ZJI-96807559-i-620008-m7     Noreen Lafflina  Very Negative   
4  DDU-69451719-O-176482-Fm  Toma Van der Beken  Very Positive   

             reason         city           state      channel response_time  \
0  Billing Question      Detroit        Michigan  Call-Center    Within SLA   
1    Service Outage  Spartanburg  South Carolina      Chatbot    Within SLA   
2  Billing Question  Gainesville         Florida  Call-Center     Above SLA   
3  Billing Question     Portland          Oregon      Chatbot    Within SLA   
4          Payments   Fort Wayne         Indiana  Call-Center    Within SLA   

      call_center  csat_score       date  call_duration  
0  Los Angeles/CA           7 2020-10-29             17  
1    Baltimore/MD           3 2020-10-05             23  
2  Los Angeles/CA           4 2020-10-04             45  
3  Los Angeles/CA           1 2020-10-17             12  
4  Los Angeles/CA           9 2020-10-17             23

### 6). Split Column (call_center)

Call_center field has city name and state name. we will seperate in and place it in two diffent fields with their respective name.

In [208]:
# split call_center field into two fields

data2[['call_center_city', 'call_center_state']] = data2['call_center'].str.split('/', expand = True)

In [209]:
# let's see the result

data2.head()

id       customer_name            reason  \
0  DKK-57076809-w-055481-fU    Analise Gairdner  Billing Question   
1  QGK-72219678-w-102139-KY    Crichton Kidsley    Service Outage   
2  GYJ-30025932-A-023015-LD   Averill Brundrett  Billing Question   
3  ZJI-96807559-i-620008-m7     Noreen Lafflina  Billing Question   
4  DDU-69451719-O-176482-Fm  Toma Van der Beken          Payments   

       sentiment      channel         city           state response_time  \
0        Neutral  Call-Center      Detroit        Michigan    Within SLA   
1  Very Positive      Chatbot  Spartanburg  South Carolina    Within SLA   
2       Negative  Call-Center  Gainesville         Florida     Above SLA   
3  Very Negative      Chatbot     Portland          Oregon    Within SLA   
4  Very Positive  Call-Center   Fort Wayne         Indiana    Within SLA   

   call_duration  csat_score   new_date      time        day     month  \
0             17           7 2022-03-01  18:08:23    Tuesday     March   
1             23           3 2022-12-28  11:04:02  Wednesday  December   
2             45           4 2022-02-23  15:18:27  Wednesday  February   
3             12           1 2022-07-11  12:34:41     Monday      July   
4             23           9 2022-08-05  13:36:01     Friday    August   

  call_center_city call_center_state     call_center  
0      Los Angeles                CA  Los Angeles/CA  
1        Baltimore                MD    Baltimore/MD  
2      Los Angeles                CA  Los Angeles/CA  
3      Los Angeles                CA  Los Angeles/CA  
4      Los Angeles                CA  Los Angeles/CA

# B). Data Processing

### a). Random Date and Time

In this dummy dataset we have date field which only include october month of 2020 and it doesn't have time also. So, for this project purpose I am adding new date field which will include only business days from jan,2022 to dec,2022 and time from 10:00 am to 8:00 pm.

In [191]:
# create business date (monday to friday) between jan,2022 to dec,2022. 

date_range = pd.date_range(start='2022-01-01', end= '2022-12-31', freq= 'B')

In [192]:
# add random date to dataframe.

data2['new_date'] = np.random.choice(date_range, size = len(data2), replace = True)

In [193]:
# create time range between 10:00 am to 8:00 pm

time_range = pd.date_range(start = '2022-01-01 10:00:00', end= '2022-01-01 20:00:00', freq='s').strftime('%H:%M:%S')

In [194]:
# add random time to datafrome

data2['time'] = np.random.choice(time_range, len(data2), replace = True)

In [195]:
# result

data2[['new_date','time']].head()

new_date      time
0 2022-03-01  18:08:23
1 2022-12-28  11:04:02
2 2022-02-23  15:18:27
3 2022-07-11  12:34:41
4 2022-08-05  13:36:01

### b). Day and Month name

We will extract day name and month name from new_date field.

In [196]:
# Generate day name

data2['day'] = data2['new_date'].dt.day_name()

In [197]:
# Generate month name

data2['month'] = data2['new_date'].dt.month_name()

In [210]:
data2.head(3)

id      customer_name            reason  \
0  DKK-57076809-w-055481-fU   Analise Gairdner  Billing Question   
1  QGK-72219678-w-102139-KY   Crichton Kidsley    Service Outage   
2  GYJ-30025932-A-023015-LD  Averill Brundrett  Billing Question   

       sentiment      channel         city           state response_time  \
0        Neutral  Call-Center      Detroit        Michigan    Within SLA   
1  Very Positive      Chatbot  Spartanburg  South Carolina    Within SLA   
2       Negative  Call-Center  Gainesville         Florida     Above SLA   

   call_duration  csat_score   new_date      time        day     month  \
0             17           7 2022-03-01  18:08:23    Tuesday     March   
1             23           3 2022-12-28  11:04:02  Wednesday  December   
2             45           4 2022-02-23  15:18:27  Wednesday  February   

  call_center_city call_center_state     call_center  
0      Los Angeles                CA  Los Angeles/CA  
1        Baltimore                MD    Baltimore/MD  
2      Los Angeles                CA  Los Angeles/CA

### c). Drop Irrelevent  Columns

We will drop date and call_center column because we have updated it.

In [214]:
data2.drop(columns = ['date', 'call_center'], axis = 1, inplace = True)

### d). Rearrange Columns

In [216]:
data2 = data2.reindex(columns = ['id', 'customer_name', 'reason', 'sentiment','channel', 'city', 'state', 'response_time', 
                       'call_duration','csat_score', 'new_date', 'time','day','month', 'call_center_city', 
                       'call_center_state'])

# C). Cleaned Data

let's see our cleaned dataset

In [221]:
data2.head()

id       customer_name            reason  \
0  DKK-57076809-w-055481-fU    Analise Gairdner  Billing Question   
1  QGK-72219678-w-102139-KY    Crichton Kidsley    Service Outage   
2  GYJ-30025932-A-023015-LD   Averill Brundrett  Billing Question   
3  ZJI-96807559-i-620008-m7     Noreen Lafflina  Billing Question   
4  DDU-69451719-O-176482-Fm  Toma Van der Beken          Payments   

       sentiment      channel         city           state response_time  \
0        Neutral  Call-Center      Detroit        Michigan    Within SLA   
1  Very Positive      Chatbot  Spartanburg  South Carolina    Within SLA   
2       Negative  Call-Center  Gainesville         Florida     Above SLA   
3  Very Negative      Chatbot     Portland          Oregon    Within SLA   
4  Very Positive  Call-Center   Fort Wayne         Indiana    Within SLA   

   call_duration  csat_score   new_date      time        day     month  \
0             17           7 2022-03-01  18:08:23    Tuesday     March   
1             23           3 2022-12-28  11:04:02  Wednesday  December   
2             45           4 2022-02-23  15:18:27  Wednesday  February   
3             12           1 2022-07-11  12:34:41     Monday      July   
4             23           9 2022-08-05  13:36:01     Friday    August   

  call_center_city call_center_state  
0      Los Angeles                CA  
1        Baltimore                MD  
2      Los Angeles                CA  
3      Los Angeles                CA  
4      Los Angeles                CA

# D). File Export

Now, our dataset is cleaned and ready for further analysis. I am exporting in csv and xlsx format.

In [220]:
# export this dataset in csv and xlsx format.

data2.to_csv('customer_data.csv', index = False)
data2.to_excel('customer_data.xlsx', index = False)